student-os/scripts/materials_convert.py

```
@@ -461,6 +481,129 @@ def get_mineru_client(ctx: ConversionContext) -> Any:
+    except json.JSONDecodeError as exc:
+        raise RuntimeError(f"MinerU v1 Agent API returned non-JSON response: {raw[:200]}") from exc
+
+
+def http_put_file(url: str, source_file: Path, *, timeout: int = 300) -> None:
+    data = source_file.read_bytes()
+    request = urllib.request.Request(
+        url,
+        data=data,
+        method="PUT",
+        headers={"Content-Type": "application/octet-stream", "Content-Length": str(len(data))},
+    )
```

Ruff (0.16.1)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)